fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/CustomItemSetting.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.item.v1;
+
+import java.util.function.Supplier;
+
+import org.jetbrains.annotations.ApiStatus;
+
+import net.minecraft.item.Item;
+
+import net.fabricmc.fabric.impl.item.CustomItemSettingImpl;
+
+/**
+ * A type of setting that can be passed to item constructors.
+ * This feature can be used by mods to add non-standard settings
+ * to items in a way that is compatible with other mods that add
+ * settings to items.
+ *
+ * <p>Values of this setting can be retrieved from an item using {@link CustomItemSetting#getValue(Item)}
+ *
+ * <p>Users that wish to expose a custom setting for use in other mods should do so by exposing
+ * the CustomItemSetting instance.
+ *
+ * <pre>{@code
+ * public static final CustomItemSetting<String> CUSTOM_TOOLTIP = CustomItemSetting.create(() -> null);
+ *
+ * @Override
+ * public void onInitializeClient() {
+ *     ItemTooltipCallback.EVENT.register((stack, context, lines) -> {
+ *         String tooltip = CUSTOM_TOOLTIP.getValue(stack.getItem());
+ *
+ *         if (tooltip != null) {
+ *             lines.add(new LiteralText(tooltip));
+ *         }
+ *     }
+ * }}</pre>
+ *
+ * <p>Do not implement. Use {@link CustomItemSetting#create(Supplier)} to retrieve an instance.
+ *
+ * @param <T> the type of the setting to be attached
+ */
+@ApiStatus.NonExtendable
```

Comment:
Why is this an interface if it shouldn't be implemented by anything else than the FAPI implementation? It think it would make more sense to have it as a normal class.

Reply:
See the history. If it's a class, implementation details leak. I think interfaces should always be the default tbh.

Reply:
Nope. In jdk's libraries, they always hide implementation details by using package-private methods. Consequently they have some package access, like JavaLangAccess etc. to access those package private methods. Imo using interface is more dangerous because people can intentionally create unsuitable implementations and break stuff even though you have that non extendable annotation there.

Reply:
I would say it is more readable to only have the relevant methods in the public API, and keep the rest under the `impl` package. I don't think we want `FabricItemApiAccess` :P

Reply:
I can tolerate another choice, this is not very important. It's just how I think about APIs.

Reply:
Another trick is like:
In the impl package utility class, you hold a `Setter` functional interface instance that has `set(CustomItemSettings type, Item.Settings settings, T value)`. Then in the static block of `CustomItemSettings`, you set that setter to a method reference to the package-private `set` function in `CustomItemSettings`.

Since the `set` in `CustomItemSettings` is not exposed, when you want to set, you will call the impl class's Setter instead. Imo this is a better way.

Reply:
Nice one. 😄

Reply:
Actually, this is exactly how java's sharedsecrets/jla access etc. works. in that impl class, you might need a check in the setter's getter method to ensure that the setter is there (i.e. `CustomItemSettings` is loaded) when the static setter field is used. I think that should be safe as when you call that setter, most likely you already have a `CustomItemSettings` and things should be already safe.